File: problems/java/RateLimiter/src/limiters/TokenBucket.java
package limiters;

import java.util.concurrent.ConcurrentHashMap;

public class TokenBucket implements RateLimiter{

    private final int capacity;
    private final int refillRate;

    // user Id -> 5 req 1 sec
    // 1 -> 5 req 1sec
    // 2 -> 5 req 1sec
    private ConcurrentHashMap<String, Bucket> buckets;

    public TokenBucket(int capacity, int refillRate) {
        this.capacity = capacity;
        this.refillRate = refillRate;
        buckets = new ConcurrentHashMap<>();
    }

    @Override
    public boolean allowRequest(String key) {
        Bucket bucket = buckets.computeIfAbsent(key, k -> new Bucket(capacity, refillRate));
        return bucket.tryConsume();
    }

    private static class Bucket {
        private final int capacity; // 10
        private final int refillRate; // 5
        private int tokens;
        private long lastRefillTime;

        public Bucket(int capacity, int refillRate) {
            this.capacity = capacity;
            this.refillRate = refillRate;
            this.tokens = capacity;
            lastRefillTime = System.currentTimeMillis();
        }

        public synchronized boolean tryConsume() {
            refill();
            if (tokens > 0) {
                tokens--;
                return true;
            }
            return false;
        }

        // 1 sec 5 req
        // 10 sec -> 10 * 5 = 50
        // cap = 5
        private void refill() {
            long now = System.currentTimeMillis();
            double timeElapsedInSeconds = (double) (now - lastRefillTime) / 1000;
            int tokenToAdd = (int)(timeElapsedInSeconds * refillRate);
            System.out.println(now+"----"+lastRefillTime+"---"+timeElapsedInSeconds+"---"+refillRate+"---"+tokenToAdd);
            tokens = Math.min(capacity, tokens + tokenToAdd);
            if (tokenToAdd > 0) lastRefillTime = now;
        }
    }
}

File: problems/python/RateLimiter/src/limiters/token_bucket.py
"""Token bucket rate limiter."""
import threading
import time
from typing import Dict


class TokenBucket:
    def __init__(self, capacity: int, refill_rate: int) -> None:
        self._capacity = capacity
        self._refill_rate = refill_rate
        self._buckets: Dict[str, "TokenBucket._Bucket"] = {}
        self._map_lock = threading.Lock()

    def allow_request(self, key: str) -> bool:
        with self._map_lock:
            if key not in self._buckets:
                self._buckets[key] = TokenBucket._Bucket(
                    self._capacity, self._refill_rate
                )
            bucket = self._buckets[key]
        return bucket.try_consume()

    class _Bucket:
        def __init__(self, capacity: int, refill_rate: int) -> None:
            self._capacity = capacity
            self._refill_rate = refill_rate
            self._tokens = capacity
            self._last_refill_time = time.time() * 1000
            self._lock = threading.Lock()

        def try_consume(self) -> bool:
            with self._lock:
                self._refill()
                if self._tokens > 0:
                    self._tokens -= 1
                    return True
                return False

        def _refill(self) -> None:
            now = time.time() * 1000
            elapsed_sec = (now - self._last_refill_time) / 1000.0
            token_to_add = int(elapsed_sec * self._refill_rate)
            self._tokens = min(self._capacity, self._tokens + token_to_add)
            if token_to_add > 0:
                self._last_refill_time = now
